File: python/Intervals/non_overlapping_intervals.py
from typing import List

class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        """
        Find the minimum number of intervals to remove to make the rest non-overlapping.
        
        Approach:
        - Sort the intervals by their start time (default sort).
        - Iterate through the intervals, keeping track of the end of the previous interval.
        - If the current interval starts after or at the end of the previous, update prevEnd.
        - If the current interval overlaps, increment the removal count and update prevEnd to the minimum end (greedy choice to keep the interval with the earliest end).
        
        Time Complexity: O(N log N), where N is the number of intervals
        - Sorting takes O(N log N)
        - The iteration is O(N)
        
        Space Complexity: O(1)
        - Only a constant amount of extra space is used
        
        Args:
            intervals: List[List[int]] - List of intervals
        Returns:
            int - Minimum number of intervals to remove
        """
        intervals.sort()

        res = 0
        prevEnd = intervals[0][1]

        for start, end in intervals[1:]:
            if start >= prevEnd:
                # No overlap, move prevEnd forward
                prevEnd = end
            else:
                # Overlap, increment removal count and keep the interval with the earliest end
                res += 1
                prevEnd = min(end, prevEnd)
        return res

# -------------------
# Test Cases
# -------------------
def test_erase_overlap_intervals():
    s = Solution()
    # Test case 1: Overlapping intervals
    assert s.eraseOverlapIntervals([[1,2],[2,3],[3,4],[1,3]]) == 1, "Failed on basic overlap"
    # Test case 2: No overlap
    assert s.eraseOverlapIntervals([[1,2],[2,3],[3,4]]) == 0, "Failed on no overlap"
    # Test case 3: All overlap
    assert s.eraseOverlapIntervals([[1,100],[1,11],[11,22],[2,12]]) == 2, "Failed on all overlap"
    # Test case 4: Single interval
    assert s.eraseOverlapIntervals([[1,2]]) == 0, "Failed on single interval"
    # Test case 5: Nested intervals
    assert s.eraseOverlapIntervals([[1,10],[2,3],[3,4],[4,5]]) == 1, "Failed on nested intervals"
    # Test case 6: Unsorted input
    assert s.eraseOverlapIntervals([[5,6],[1,3],[2,4]]) == 1, "Failed on unsorted input"
    print("All test cases passed!")

if __name__ == "__main__":
    test_erase_overlap_intervals()

File: C++/Intervals/non_overlapping_intervals.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * Find the minimum number of intervals to remove to make the rest non-overlapping.
     *
     * Time Complexity: O(N log N), where N is the number of intervals
     *   - Sorting takes O(N log N)
     *   - Iteration is O(N)
     *
     * Space Complexity: O(1)
     *   - Only a constant number of variables are used
     */
    int eraseOverlapIntervals(vector<vector<int>>& intervals) {
        // Sort intervals by their start time
        sort(intervals.begin(), intervals.end());

        int res = 0;  // Counter for removals
        int prevEnd = intervals[0][1];  // End of the previous interval

        for (int i = 1; i < intervals.size(); i++) {
            int start = intervals[i][0];
            int end = intervals[i][1];

            if (start >= prevEnd) {
                // No overlap: move prevEnd forward
                prevEnd = end;
            } else {
                // Overlap: increment removal count and keep interval with smallest end
                res++;
                prevEnd = min(prevEnd, end);
            }
        }

        return res;
    }
};

/*
===========================================================
Line-by-line C++ syntax explanation:
===========================================================

#include <bits/stdc++.h>
    → Includes all standard C++ libraries (like vector, algorithm, iostream, etc.).

using namespace std;
    → Allows you to use standard library names without prefixing them with std::.

class Solution {
public:
    → Defines a class named Solution with public methods.

int eraseOverlapIntervals(vector<vector<int>>& intervals) {
    → Defines a method that:
       - Takes a reference to a vector of vectors of integers (`intervals`)
       - Returns an integer

sort(intervals.begin(), intervals.end());
    → Sorts the intervals in increasing order of their start time.

int res = 0;
    → Initialize a counter for the number of intervals to remove.

int prevEnd = intervals[0][1];
    → Initialize `prevEnd` as the end of the first interval.

for (int i = 1; i < intervals.size(); i++) {
    → Loop through the intervals starting from the second one.

int start = intervals[i][0];
int end = intervals[i][1];
    → Get the start and end of the current interval.

if (start >= prevEnd) {
    → If there is no overlap:
prevEnd = end;
    → Update `prevEnd` to the end of the current interval.

else {
    → Overlap detected:
res++;
    → Increment the count of removals.

prevEnd = min(prevEnd, end);
    → Keep the interval with the smaller end time (greedy choice).

return res;
    → Return the minimum number of intervals to remove.

===========================================================
Time Complexity: O(N log N) because of sorting.
Space Complexity: O(1) since no extra space proportional to N is used.
===========================================================
*/

